packages/swingset/src/get-nav-tree.ts

```diff
@@ -1,49 +1,95 @@
-import { ComponentEntity, DocsEntity } from './types.js'
+import {
+  ComponentEntity,
+  DocsEntity,
+  ComponentNode,
+  CategoryNode,
+} from './types.js'
 
 export function getNavigationTree(entities: (ComponentEntity | DocsEntity)[]) {
-  let result: Record<
-    string,
-    Pick<ComponentEntity, 'title' | 'slug' | 'componentPath' | 'children'>[]
-  > = {}
+
 
   const componentEntities = entities.filter(
     (entity) => entity.__type === 'component'
   ) as ComponentEntity[]
 
-  const componentEntitiesWithChildren = componentEntities.map((entity) => {
-    if (entity.isNested) return entity
+  const componentEntitiesWithChildren = componentEntities.map(
+    (componentEntity) => {
+      if (componentEntity.isNested) return componentEntity
+
+      componentEntity.children = componentEntities.filter(
+        (childEntity) =>
+          childEntity.isNested &&
+          childEntity.componentPath === componentEntity.componentPath
+      )
 
-    entity.children = componentEntities.filter(
-      (childEntity) =>
-        childEntity.isNested &&
-        childEntity.componentPath === entity.componentPath
-    )
+      return componentEntity
+    }
+  )
+
+  //TODO: Account for duplicate folder names [category]/[folder]
+  const categories = new Map<CategoryNode['title'], CategoryNode>()
 
-    return entity
-  })
 
   // bucket components into categories, nested documents are categorized under their component's path
   for (const entity of componentEntitiesWithChildren) {
     if (entity.isNested) continue
 
-    //TODO: Handle Default Category
-    const category = entity.navigationData?.category!
-
-
-
-    result[category] ||= []
-
-
-  
-    result[category].push({
+    const entityData: ComponentNode = {
+      type: 'component',
       title: entity.title,
       slug: entity.slug,
       componentPath: entity.componentPath,
       children: entity.children,
-    })
+    }
+
+  
+    const categoryTitle = entity.navigationData?.category || 'default'
+    const folderTitle = entity.navigationData?.folder
+    const hasFolder = !!folderTitle
+
+    const storedCategory =
+      categories.has(categoryTitle) && categories.get(categoryTitle)
+
+     
+
+    if (!!storedCategory) {
```

Comment:
I would potentially split out the category creation (if not exists) from the folder handling. I think the flow would look like (in plain english):

```
check if category is stored

if category is not stored, create the category object

then, handle folder logic
```

This dedupes some of the folder logic between these two branches.

(optional) I also wonder if it'd be easier to build up the category map and folder map separately, and then after we're done iterating over all of the entities we can make sure each folder gets added to its proper category. that way we don't need to check each of a category's children for the correct folder when we're trying to add an element to it.